### ISSUE TITLE
Proposal: refactor LintRule trait, message, hint

### ISSUE THREAD

**Issue by magurotuna** [denoland/deno_lint#457]
I propose we do refactoring `LintRule` trait into:

```rust
pub trait LintRule {
  type Message;
  type Hint;

  const CODE: &'static str;
  const TAGS: &'static [&'static str];
  const DOCS: &'static str;

  fn new() -> Box<Self>
  where
    Self: 'static + Sized;

  fn code(&self) -> &'static str {
    Self::CODE
  }

  fn tags(&self) -> &'static [&'static str] {
    Self::TAGS
  }

  fn docs(&self) -> &'static str {
    Self::DOCS
  }

  fn lint_program(
    &self,
    context: &mut Context,
    program: &swc_ecmascript::ast::Program,
  );
}
```

After this refactoring, trait implementation will look like:

```rust
use derive_more::Display; // comes from `derive_more` crate

#[derive(Display)]
pub enum FooRuleMessage {
  #[display(fmt = "ooops!")]
  Unexpected,
}

#[derive(Display)]
pub enum FooRuleHint {
  #[display(fmt = "some hint")]
  Bar,
  #[display(fmt = "This is dummy hint with two strings! {} {}", _0, _1)]
  Dummy(String, String),
}

impl LintRule for FooRule {
  type Message = FooRuleMessage;
  type Hint = FooRuleHint;

  const CODE: &'static str = "foo_rule";
  const TAGS: &'static [&'static str] = &["recommended"];
  const DOCS: &'static str = "some documents";

  fn new() -> Box<Self> {
    Box::new(FooRule)
  }

  fn lint_program(
    &self,
    context: &mut Context,
    program: &swc_ecmascript::ast::Program,
  ) {
    // snip
  }
}

```

## Motivation 

There are two reasons why I want to refactor. This refactor will allow us to:

1. avoid duplicate "code" definitions
2. get statically typed messages & hints that are associated with `LintRule` type

### 1. avoid duplicate "code" definitions

We already have `code` method on `LintRule` trait but it is not used when error messages are generated. I mean, to take `eqeqeq` rule for example, a diagnostic is added like this:

```rust
      self.context.add_diagnostic_with_hint(
        bin_expr.span,
        "eqeqeq",
        message,
        hint,
      )
```

the rule code (`"eqeqeq"`) is directly specified here, even though `Eqeqeq` struct has `code` method that returns `"eqeqeq"`.  This is not only redundant but also could cause a bug due to typo. Rule code should be defined only once.

If we make `LintRule` trait have associated const `CODE`, then we will be able to rewrite the `eqeqeq` example to:

```rust
      self.context.add_diagnostic_with_hint(
        bin_expr.span,
        Eqeqeq::CODE,
        message,
        hint,
      )
```

which seems quite nice to me.

### 2. get statically typed messages & hints

We are in the middle of switching to `assert_lint_err!` macro. This macro allows us to make sure messages and hints are equal to what we expect. 
This is so powerful, but how test cases look like become a kind of mess, for example: https://github.com/denoland/deno_lint/blob/v0.2.5/src/rules/camelcase.rs#L625-L652
What we really want to assert is not necessarily message / hint *as string*, but what *kind* of message / hint it is.

Actually, in ESLint, the error messages are managed by message IDs (e.g. `unexpected`) and corresponding format strings (e.g. `"Expected '{{expectedOperator}}' and instead saw '{{actualOperator}}'."`).
https://github.com/eslint/eslint/blob/v7.12.1/lib/rules/eqeqeq.js#L63-L65
If we imitate it in Rust, thet's where `enum` comes in. It will allow us to acquire statically typed messages and hints.

If we create such enum, then we will be able to write a test like:

```rust
    assert_lint_err! {
      FooRule,
      r#"function foo_bar(){}"#: [
            {
              col: 9,
              message: Unexpected,
              hint: Dummy("foo_bar", "baz"),
            }
          ],
```

This looks even more readable to me than it is now.

**Comment by RDambrosio016** [comment]
That will not work because trait objects do not support associated constants

**Comment by magurotuna** [comment]
ah that's right... too bad 😢

**Comment by magurotuna** [comment]
We cannot make it work, but I think declaring messages and hints as enum with `derive_more` still has room for consideration.

**Comment by bartlomieju** [comment]
@magurotuna I thought about it in #424 but as @RDambrosio016 points out it's not doable due to trait object limitations. I'm open to refactors messages and hints as suggested.

**Comment by magurotuna** [comment]
@bartlomieju 

> I thought about it in #424 but as @RDambrosio016 points out it's not doable due to trait object limitations. 

Actually I didn't think of #424 as a problem with trait object limitation - now I get it.

> I'm open to refactors messages and hints as suggested.

OK. Don't you mind my adding `derive_more` with `display` feature as a new dependency?

```
derive_more = { version = "0.99.11", features = ["display"] }
```

**Comment by bartlomieju** [comment]
@magurotuna seems like a very elegant solution - let's try on a few rules with variable messages and see how it works

**Comment by bartlomieju** [comment]
@magurotuna one more thing; in #424 @RDambrosio016 proposed to implement rules [not as traits but regular objects](https://github.com/denoland/deno_lint/issues/424#issuecomment-716723621) - maybe it's worth exploring this approach

**Comment by magurotuna** [comment]
@bartlomieju 

> let's try on a few rules with variable messages and see how it works

okay, I will try this approach on one rule tomorrow.

> in #424 @RDambrosio016 proposed to implement rules not as traits but regular objects - maybe it's worth exploring this approach

sounds interesting... I also will dig deep into that approach.

**Comment by RDambrosio016** [comment]
That's not exactly what i meant, i meant to just separate rule metadata and rule implementation into two separate traits like rslint does.
e.g. [`Rule`](https://docs.rs/rslint_core/0.2.2/rslint_core/trait.Rule.html) which has name, group, and docs
and [`CstRule`](https://docs.rs/rslint_core/0.2.2/rslint_core/trait.CstRule.html) which has the actual implementation functions

**Comment by magurotuna** [comment]
Closing as this should be considered further and is not an urgent issue. I'll open a new issue when the time comes.
